Add Delayed State to Drag and Drop

Phipson opened the pull request:
#### 8816e1930e888a014fac89d18dd8c42230263922
<pre>
Add Delayed State to Drag and Drop
<a href="https://bugs.webkit.org/show_bug.cgi?id=291284">https://bugs.webkit.org/show_bug.cgi?id=291284</a>
<a href="https://rdar.apple.com/148843740">rdar://148843740</a>

Reviewed by Wenson Hsieh and Ada Chan.

Changes the return type of `tryToBeginDragAtPoint` and arg type of `didHandleDragStartRequest` and
`didHandleAdditionalDragItemsRequest` from a bool to an enum. This is intended to handle potential cases (in
subsequent PRs) where the web or model process needs to perform some asynchronous actions before completing
the request. As a result, we need to add a `delayed` state to the request, telling the UI Process that the
drag and drop start request has not been completed but has not ended or begun yet either.

* Source/WebCore/page/DragActions.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDragStart):
(WebKit::WebPage::requestAdditionalItemsForDragSession):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _requestStartDataInteraction:globalPosition:]):

Canonical link: <a href="https://commits.webkit.org/293663@main">https://commits.webkit.org/293663@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79e82ca6e3a9a42fccb6118cc19ba4c864a76b4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99375 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9275 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104506 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49976 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19313 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27458 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75634 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32733 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102382 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14685 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89726 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55993 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/98848 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14479 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7709 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49336 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7796 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106863 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26489 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19437 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84595 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26851 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85930 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84108 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21380 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28780 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6470 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20232 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26429 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31630 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26249 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29562 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27816 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1126,7 +1126,7 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
     localMainFrame->eventHandler().dragSourceEndedAt(event, { }, MayExtendDragSession::Yes);
 
-    bool didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
+    DragStartRequestResult didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
```
Curious whether it's worth sending a message to the UI process about the delay. Could we just `Messages::WebPageProxy::DidHandleAdditionalDragItemsRequest` when we get a non-delayed result? Or maybe have `tryToBeginDragAtPoint` return asynchronously? Would be curious in what @whsieh thinks.

achan00 commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -364,7 +364,7 @@ class EventHandler final : public CanMakeCheckedPtr<EventHandler> {
 
 #if PLATFORM(IOS_FAMILY) && ENABLE(DRAG_SUPPORT)
-    WEBCORE_EXPORT bool tryToBeginDragAtPoint(const IntPoint& clientPosition, const IntPoint& globalPosition);
+    WEBCORE_EXPORT bool tryToBeginDragAtPoint(const IntPoint& clientPosition, const IntPoint& globalPosition, CompletionHandler<void(bool)>&&);
```
Since we are returning the boolean result in a completion handler, the return value of this method should be `void`:

```
WEBCORE_EXPORT void tryToBeginDragAtPoint(const IntPoint& clientPosition, const IntPoint& globalPosition, CompletionHandler<void(bool)>&&);
```

achan00 replied:
Actually please ignore this comment and focus on the last one: https://github.com/WebKit/WebKit/pull/43812#discussion_r2036447315

achan00 commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -811,11 +813,13 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
 
     auto subframe = dynamicDowncast<LocalFrame>(subframeForHitTestResult(hitTestedMouseEvent));
-    if (subframe && subframe->eventHandler().tryToBeginDragAtPoint(adjustedClientPosition, adjustedGlobalPosition))
+    if (subframe && subframe->eventHandler().tryToBeginDragAtPoint(adjustedClientPosition, adjustedGlobalPosition, WTFMove(completionHandler)))
```
It's a little tricky to preserve the previous logic here. One thing we can do is to capture the logic after this (lines starting in line 819) in a closure (or in another method). In the completion handler of the recursive `tryToBeginDragAtPoint`, if the result is true, call the original completion handler with `true`, otherwise call the closure. And if there's no `subframe`, call the closure too.

It's a bit messy, and maybe @whsieh has a better suggestion here.

achan00 replied:
Actually please ignore this comment and focus on the last one: https://github.com/WebKit/WebKit/pull/43812#discussion_r2036447315

achan00 commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1108,8 +1108,10 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
-    bool didStart = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
-    send(Messages::WebPageProxy::DidHandleDragStartRequest(didStart));
+
+    localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition, [this](bool didStart) {
```
Should probably capture `weakThis = WeakPtr { *this }` and set it to a `RefPtr` and check for null before using it inside the block.

achan00 replied:
Actually please ignore this comment and focus on the last one: https://github.com/WebKit/WebKit/pull/43812#discussion_r2036447315

achan00 commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1126,8 +1128,9 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
-    bool didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
-    send(Messages::WebPageProxy::DidHandleAdditionalDragItemsRequest(didHandleDrag));
+    localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition, [this](bool didStart) {
+        send(Messages::WebPageProxy::DidHandleAdditionalDragItemsRequest(didStart));
```
Ditto

achan00 replied:
Actually please ignore this comment and focus on the last one: https://github.com/WebKit/WebKit/pull/43812#discussion_r2036447315

achan00 commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -1920,13 +1920,16 @@ - (void)updateLayoutIgnorePendingStyleSheets
 #if ENABLE(DRAG_SUPPORT)
 
-- (BOOL)_requestStartDataInteraction:(CGPoint)clientPosition globalPosition:(CGPoint)globalPosition
+- (BOOL)_requestStartDataInteraction:(CGPoint)clientPosition globalPosition:(CGPoint)globalPosition completionHandler:(void (^)(bool)) completionHandler
```
Nit: remove the space before `completionHandler` in `(void (^)(bool)) completionHandler`.

achan00 replied:
Actually - I don't think we can change the signature of this `WebView` API. Sorry I didn't foresee this. This makes it difficult for `tryToBeginDragAtPoint` to use a completion handler. 

The best option I can think of now is to bring back part of your previous version of the PR, but revert the `WebPageProxy`'s `DidHandleDragStartRequest` and `DidHandleAdditionalDragItemsRequest` messages to take in a boolean as before. If `tryToBeginDragAtPoint` returns a delayed result, then send those messages after   the model process finishes the drag preparation work.

achan00 commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -811,14 +811,14 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
 
     handleMousePressEvent(syntheticMousePressEvent);
-    bool handledDrag = m_mouseDownMayStartDrag && handleMouseDraggedEvent(hitTestedMouseEvent, DontCheckDragHysteresis);
+    DragStartRequestResult handledDrag = m_mouseDownMayStartDrag && handleMouseDraggedEvent(hitTestedMouseEvent, DontCheckDragHysteresis) ? DragStartRequestResult::Ended : DragStartRequestResult::Started;
```
Should this be:
```
DragStartRequestResult handledDrag = m_mouseDownMayStartDrag && handleMouseDraggedEvent(hitTestedMouseEvent, DontCheckDragHysteresis) ? DragStartRequestResult::Started : DragStartRequestResult::Ended;
```

achan00 commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -1925,8 +1925,9 @@ - (BOOL)_requestStartDataInteraction:(CGPoint)clientPosition globalPosition:(CGP
     if (!localMainFrame)
-        return NO;
-    return localMainFrame->eventHandler().tryToBeginDragAtPoint(WebCore::IntPoint(clientPosition), WebCore::IntPoint(globalPosition));
+        return false;
```
Nit: `return NO;`

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1108,8 +1108,11 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
-    bool didStart = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
-    send(Messages::WebPageProxy::DidHandleDragStartRequest(didStart));
+
+    WebCore::DragStartRequestResult didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
```
Nit - might be a bit cleaner like this

```suggestion
    auto didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
```

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1126,8 +1129,10 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
 
-    bool didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
-    send(Messages::WebPageProxy::DidHandleAdditionalDragItemsRequest(didHandleDrag));
+    WebCore::DragStartRequestResult didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
```
Ditto.

```suggestion
    auto didHandleDrag = localMainFrame->eventHandler().tryToBeginDragAtPoint(clientPosition, globalPosition);
```

Phipson committed:
Add Delayed State to Drag and Drop
https://bugs.webkit.org/show_bug.cgi?id=291284
rdar://148843740

Reviewed by Wenson Hsieh and Ada Chan.

Changes the return type of `tryToBeginDragAtPoint` and arg type of `didHandleDragStartRequest` and
`didHandleAdditionalDragItemsRequest` from a bool to an enum. This is intended to handle potential cases (in
subsequent PRs) where the web or model process needs to perform some asynchronous actions before completing
the request. As a result, we need to add a `delayed` state to the request, telling the UI Process that the
drag and drop start request has not been completed but has not ended or begun yet either.

* Source/WebCore/page/DragActions.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDragStart):
(WebKit::WebPage::requestAdditionalItemsForDragSession):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _requestStartDataInteraction:globalPosition:]):

Canonical link: https://commits.webkit.org/293663@main
